maintainers/scripts/haskell/hydra-report.hs

```diff
@@ -132,30 +130,104 @@ getBuildReports = runReq defaultHttpConfig do
 hydraEvalParams = ["-I", ".", "pkgs/top-level/release-haskell.nix"]
+
 handlesCommand :: FilePath
 handlesCommand = "nix-instantiate"
+
 handlesParams :: [String]
 handlesParams = ["--eval", "--strict", "--json", "-"]
+
 handlesExpression :: String
 handlesExpression = "with import ./. {}; with lib; zipAttrsWith (_: builtins.head) (mapAttrsToList (_: v: if v ? github then { \"${v.email}\" = v.github; } else {}) (import maintainers/maintainer-list.nix))"
 
-newtype Maintainers = Maintainers {maintainers :: Maybe Text} deriving (Generic, ToJSON, FromJSON)
-
+newtype Maintainers = Maintainers {maintainers :: Text}
+  deriving stock (Generic)
+  deriving anyclass (FromJSON, ToJSON)
+
+-- | This is a 'Map' from Hydra job name to maintainer email addresses.
+--
+-- It has values similar to the following:
+--
+-- @@
+--  fromList
+--    [ ("arion.aarch64-linux", Maintainers "robert@example.com")
+--    , ("bench.x86_64-linux", Maintainers "")
+--    , ("conduit.x86_64-linux", Maintainers "snoy@man.com, web@ber.com")
+--    , ("lens.x86_64-darwin", Maintainers "ek@category.com")
+--    ]
+-- @@
+--
+-- Note that Hydra jobs without maintainers will have an empty string for the
+-- maintainer list.
 type HydraJobs = Map Text Maintainers
+
+-- | Map of email addresses to GitHub handles.
+-- This is built from the file @../../maintainer-list.nix@.
+--
+-- It has values similar to the following:
+--
+-- @@
+--  fromList
+--    [ ("robert@example.com", "rob22")
+--    , ("ek@category.com", "edkm")
+--    ]
+-- @@
+type EmailToGitHubHandles = Map Text Text
+
+-- | Map of Hydra jobs to maintainer GitHub handles.
+--
+-- It has values similar to the following:
+--
+-- @@
+--  fromList
+--    [ ("arion.aarch64-linux", ["rob22"])
+--    , ("conduit.x86_64-darwin", ["snoyb", "webber"])
+--    ]
+-- @@
 type MaintainerMap = Map Text (NonEmpty Text)
 
+-- | Generate a mapping of Hydra job names to maintainer GitHub handles.
 getMaintainerMap :: IO MaintainerMap
 getMaintainerMap = do
-   hydraJobs :: HydraJobs <- get hydraEvalCommand hydraEvalParams "" "Failed to decode hydra-eval-jobs output: "
-   handlesMap :: Map Text Text <- get handlesCommand handlesParams handlesExpression "Failed to decode nix output for lookup of github handles: "
-   pure $ hydraJobs & Map.mapMaybe (nonEmpty . mapMaybe (`Map.lookup` handlesMap) . Text.splitOn ", " . fromMaybe "" . maintainers)
-  where
-   get c p i e = readProcess c p i <&> \x -> either (error . (<> " Raw:'" <> take 1000 x <> "'") . (e <>)) Prelude.id . eitherDecodeStrict' . encodeUtf8 . Text.pack $ x
+   hydraJobs :: HydraJobs <-
+      readJSONProcess hydraEvalCommand hydraEvalParams "" "Failed to decode hydra-eval-jobs output: "
+   handlesMap :: EmailToGitHubHandles <-
+      readJSONProcess handlesCommand handlesParams handlesExpression "Failed to decode nix output for lookup of github handles: "
+   pure $ Map.mapMaybe (nonEmpty . mapMaybe (`Map.lookup` handlesMap) . Text.splitOn ", " . maintainers) hydraJobs
```

Comment:
The only substantial change I made here was to change 

```haskell
newtype Maintainers = Maintainers {maintainers :: Maybe Text}
```
to
```haskell
newtype Maintainers = Maintainers {maintainers :: Text}
```

This seemed to be okay, since `hydra-eval-jobs` has a `maintainer` for every job.  For jobs without a maintainer, an empty string is used (not the value `null`).

Reply:
I see your point. But this had been a `Text` before and I specifically changed this because there were cases where this invariant did not hold. (I am not sure why and when exactly, but there was a case where a job had not maintainer field.) So I think being a bit lenient at this point is probably the more stable solution.

Reply:
Ah, yeah, I was wondering if that was the case!

I'll revert this part, make sure it still runs, and merge this PR in.

Reply:
I made this change in 4ac4ced and confirmed the script still runs, so I'll go ahead and merge this in.

Reply:
Oh, I might have figured this out.

When a package is marked as `broken`, you get an entry from `hydra-eval-jobs` that looks like this:

```json
{
  "haskell.compiler.native-bignum.ghcHEAD.aarch64-linux": {
    "error": "error: Failed to evaluate ghc-9.3.20210504: «broken»: is marked as broken"
  }
}
```